### PR TITLE
Update call hierarchy for lsp4j

### PR DESCRIPTION
**Pull request by testforstephen**
The call hierarchy protocol in LSP has updated to a new version, see https://github.com/microsoft/vscode-languageserver-node/blob/master/protocol/src/protocol.callHierarchy.proposed.ts.

It includes three new requests:
- textDocument/prepareCallHierarchy
- callHierarchy/incomingCalls
- callHierarchy/outgoingCalls

Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

### CODE REVIEW

**Comment by jonahgraham** [comment]
> Can one of the admins verify this patch?

run tests

(Note I kicked off the tests, but I am not in a position to review at the moment)

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend [hunk 1]
```diff
@@ -4850,73 +4850,102 @@ class SemanticHighlightingInformation {
...
+	CallHierarchyItem from
 
 	/**
-	 * The symbol that contains the reference.
+	 * The range at which at which the calls appears. This is relative to the caller
+	 * denoted by [`this.from`](#CallHierarchyIncomingCall.from).
 	 */
-	@NonNull
-	CallHierarchySymbol from
+	Range[] fromRanges
```
Comment (KamasamaK) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]:
```suggestion
	List<Range> fromRanges
```

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend [hunk 2]
```diff
@@ -4850,73 +4850,102 @@ class SemanticHighlightingInformation {
...
-	@NonNull
-	CallHierarchySymbol to
+	CallHierarchyItem to
 
+	/**
+	 * The range at which this item is called. This is the range relative to the caller, e.g the item
+	 * passed to [`provideCallHierarchyOutgoingCalls`](#CallHierarchyItemProvider.provideCallHierarchyOutgoingCalls)
+	 * and not [`this.to`](#CallHierarchyOutgoingCall.to).
+	 */
+	Range[] fromRanges
```
Comment (KamasamaK) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]:
```suggestion
	List<Range> fromRanges
```

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend [hunk 3]
```diff
@@ -4850,73 +4850,102 @@ class SemanticHighlightingInformation {
...
+	 * The kind of this item.
 	 */
 	@NonNull
 	SymbolKind kind
 
 	/**
-	 * URI of the document containing the symbol.
+	 * Tags for this item.
+	 */
+	SymbolTag[] tags
```
Comment (KamasamaK) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]:
```suggestion
	List<SymbolTag> tags
```

Comment (KamasamaK) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: `SymbolTag` is not defined here.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend (hunk 1 above)
Comment (testforstephen) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: fixed

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend (hunk 3 above)
Comment (testforstephen) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: SymbolTag is enum type. I saw other enums are all manually written. Is there any enum sample for Xtend definition?

Comment (testforstephen) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: Looks like enum with constructor is not supported well in Xtend. https://bugs.eclipse.org/bugs/show_bug.cgi?id=403783

Comment (KamasamaK) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: You're right, the enums for LSP seem to be manual. I must have just remembered seeing them in [DebugProtocol.xtend](https://github.com/eclipse/lsp4j/blob/master/org.eclipse.lsp4j.debug/src/main/java/org/eclipse/lsp4j/debug/DebugProtocol.xtend).

Comment (testforstephen) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: Yeah, i have a try on adding SymbolTag enum to Xtend, but it doesn't work.

Here are the problems:
- SymbolTag in LSP starts with 1, so the default enum definition doesn't work.
```java
enum SymbolTag {
  Deprecated
}
```
- @SerializedName can only specify an alias string for the enum type. But the value passed from LSP is a int number, the jsonrpc cannot convert it to the serializedName SymbolTag.
```java
enum SymbolTag {
  @SerializedName("1")
  Deprecated
}
```

**Pull request merged**